Polina auth

PtFux opened the pull request:
Сервис регистрации.
На данный момент:
- работают обе ручки
- работает мидлваре аутентификации

PtFux committed:
added auth api docs

PtFux committed:
added auth api

PtFux committed:
corrected auth politics

PtFux committed:
added view-models

добавлены модели для передачи/получения при взаимодействии с фронтендом

PtFux committed:
added validation error

PtFux committed:
rename auth project

изменено имя проекта, чтобы соответствовать гитхабу

PtFux committed:
planed behavior

PtFux committed:
renamed auth project

PtFux committed:
added parsing and valifation to AuthRegisterPost | stable version

+ added parsing and valifation to AuthRegisterPost // validation in reg.go
+ renamed project

PtFux committed:
great migration of files

PtFux committed:
corrected openapi.yaml

PtFux committed:
added instruction to logging util

PtFux committed:
changed models

- added common InitValidErrorInField
- added login model
- added common check

PtFux committed:
added python test

PtFux committed:
added parse and validation in AuthLoginPost

PtFux committed:
updated readme.md in auth

kanny4uHo commented on go.mod:
```diff
@@ -0,0 +1,5 @@
+module github.com/go-park-mail-ru/2024_2_TeamOn_Patreon/tree/polina-auth
```
модуль надо как-то попроще назвать, типа backend, project или что-то такое простое и обобщающее

kanny4uHo replied:
тогда и импорты будут не такие жесткие

PtFux replied:
Так надо же ссылкой с гита

kanny4uHo replied:
почему?

kanny4uHo replied:
обсудили, допилили

kanny4uHo commented on internal/auth/api/api_auth.go:
```diff
@@ -0,0 +1,63 @@
+	// parse models
+	var l models.Login
+	if err := json.NewDecoder(r.Body).Decode(&l); err != nil {
+		slog.Info(fmt.Sprintf("error {%v} | in %v", err, op))
```
давай сделаем свой пакет logger. там прям интерфейс Logger, его реализация, и чтобы можно было создать типа методом New его. А он пусть использует slog

kanny4uHo replied:
еще можно будет допилить методы типа Infof, чтобы тот уже внутри форматировал строку как надо с помощью fmt.Sprintf()

PtFux replied:
Оке

kanny4uHo commented on internal/auth/utils/logging.go:
```diff
@@ -0,0 +1,31 @@
+/*
+** HOW TO USE ??
+
+Предлагаю использовать при создании роутера
```
для этого есть middlewares, лучше их заюзать. А хендлеры сделать как структуры с реализацией функции обработки запроса. Структура - это чтобы туда можно было передавать аккуратно структуры логирования, бд и тому подобное

kanny4uHo commented on internal/auth/api/models/login.go:
```diff
@@ -0,0 +1,90 @@
+
+	checks := []check{
+		check{pattern: `^[a-zA-Z0-9!@#$%^&*()_+={}:|,.?]+$`,
+			msg: "пароль может содержать только буквы, цифры и спец символы: ^[a-zA-Z0-9!@#$%^&*()_+={}:|,.?]+$"},
```
это же юзеру показываться будет? давай не будем палить регулярку, по которой мы матчим)

PtFux replied:
Ой, там должен был быть только набор символов)

kanny4uHo commented on internal/auth/behavior/behavior.go:
```diff
@@ -0,0 +1,3 @@
+package behavior
```
можно здесь handlers сделать, писать логику в них

PtFux replied:
Хэндерс это другой уровень

kanny4uHo replied:
хочешь прям usecases, по чистой архитектуре?)

kanny4uHo replied:
useCase -> behavior

PtFux replied:
что-то вроде того

kanny4uHo commented on internal/auth/errors/validation_error.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+// InitValidationError - инициализация ошибки
+func InitValidationError(err string, msg string) *ValidationError {
```
NewValidationError

kanny4uHo commented on internal/auth/api/models/login.go:
```diff
@@ -0,0 +1,90 @@
+	Password string `json:"password"`
+}
+
+func (lg *Login) Validate() (bool, *er.ValidationError) {
```
в models стоит хранить чисто входные и выходные структуры. давай вынесем валидацию в behavior

kanny4uHo replied:
может даже в отдельную папку validation

PtFux replied:
Оке

PtFux replied:
Только почему в бехэйвор, валидация же внешних данных, в бехейвере должны быть идеальные

kanny4uHo replied:
потому что валидация - это бизнес логика (хотя я читал холивары на эту тему - бизнес логика или данные, но здесь валидация мега общая, в какие-то момент для нее даже понадобятся походы в бд, поэтому лучше в behavior)

PtFux replied:
а может разделим бизнес-валидацию и валидацию входных данных ? 
и когда будет потребность ходить в бд, сделаем валидатор внутри бизнес логики.

тут валидация завязана на фронтовой модели, на той структуре, которую получаем из сети. 

Если сейчас выносить эту валидацию в бизнес логику, то будет не оч красиво выглядеть) и придется маппить именно эти данные в бизнес модель, т.е. придется создавать бизнес модель в виде структуры, которая приходит с фронта, либо писать валидадатор для каждого поля, а логика валидации пароля в логине  и регистарции, например, немного отличается

kanny4uHo commented on internal/auth/api/models/login.go:
```diff
@@ -0,0 +1,90 @@
+
+	passValid, err := lg.validatePassword()
+	if err != nil || !passValid {
+		return false, err
```
есть такой пакет - github.com/pkg/errors. С его помощью можно оборачивать ошибки в какой-то контекст, или создавать новые.  и вот когда возвращаешь ошибки, лучше их заворачивать в какой-то контекст. здесь типа "failed to validate password". и обработать !passValid отдельно, потому что по сути это другая ошибка.

kanny4uHo commented on internal/auth/api/test/access.py:
```diff
@@ -0,0 +1,37 @@
+from dataclasses import dataclass
+
+# Define the URL of the server
+BASE_URL = "http://localhost:8081/"
```
фига, сильно. а тесты на го не хотим писать?)

PtFux replied:
(Да)

PtFux replied:
Попозже будут 🙏

PtFux committed:
added many logics

Есть две ручки - регистрации и авторизации.
- регистрация -
  - регистрирует пользователя
  - проставляет токен в куки
  - возвращает 200, если ок, иначе - 400
- аутентификация -
  - аутентифицирует пользователя,
  - проставляет токен в куки
  - возвращает 200, если ок, иначе - 400

Также создана общая middleware, которая аутентифицирует пользователя по куки,
  - если все ок, передает бизнес модель пользователя в контекст,
  - иначе, возвращает 401

Также создана middleware
  - для логирования
  - для передачи структуры бизнес-логики

Также создан общий модуль логирования

PtFux committed:
added logging response

PtFux committed:
auth: added docker

PtFux committed:
auth: added sending error model

PtFux committed:
auth: changed module name

kanny4uHo commented on internal/auth/behavior/behavior.go:
```diff
@@ -0,0 +1,173 @@
+	// Проверка есть ли такой username
+	// если произошла ошибка, вернуть её
+	exists, errM := b.isUserExists(username)
+	if exists == nil || *exists {
```
здесь нужно почекать ошибку сначала и вернуть, если она есть, потом уже проверки exists. флаг булевый можно сделать не указателем

kanny4uHo commented on internal/auth/behavior/behavior.go:
```diff
@@ -0,0 +1,173 @@
+	// хэширование пароля
+	hash, errM := hashPassword(password)
+	if errM != nil {
+		return "", errM
```
возвращаем ошибки, заворачивая в какой-то контекст. Я писал выше про пакет, который позволяет это делать на изи

kanny4uHo commented on internal/common/interfaces/repository.go:
```diff
@@ -0,0 +1,5 @@
+package interfaces
+
+// Repository интерфейс структуры по работе с БД
+type Repository interface {
```
в такого рода интерфейсы что планируется добавлять?

PtFux replied:
удаляем его, был для абстрактности мидлваре

kanny4uHo commented on cmd/microservices/auth/main.go:
```diff
@@ -1,23 +1,35 @@
-func main() {
-	http.HandleFunc("/auth/", authHandler)
+	// logger
+	logger.New()
```
если создавать логгер на старте каждого хендлера, ты туда сможешь передать информацию, специфичную для каждой ручки и запроса. и дописывать их как хочешь, без параметра в методе "op" или "req-id"

PtFux replied:
там, так не очень сработает, только если опять через замыкания... ну и в структуру нужно будет хэндлеры перенести

kanny4uHo replied:
вот у тебя handler начинает работать

kanny4uHo replied:
там можно разместить создания логгера

kanny4uHo commented on internal/common/buisness/models/role.go:
```diff
@@ -0,0 +1,24 @@
+
+// Именованные константы для ролей
+const (
+	Reader Role = iota + 1 // Читатель
```
константы лучше явно указать в этом случае, чтобы если кто-то добавил роль в середину, ничего не поплыло

kanny4uHo commented on internal/auth/api/login_post.go:
```diff
@@ -0,0 +1,72 @@
+)
+
+// LoginPost - ручка аутентификации
+func LoginPost(w http.ResponseWriter, r *http.Request) {
```
мб сделать этот метод на структуре Handler? у нее сделать поле behavior, туда складывать, и брать тут. Не через контекст запроса

kanny4uHo replied:
целая middleware исчезнет

PtFux replied:
так и сделаеем, но мб после 1 рк..?

kanny4uHo replied:
ок, отложим

PtFux replied:
исполненео

kanny4uHo commented on internal/auth/behavior/behavior.go:
```diff
@@ -0,0 +1,173 @@
+	exists, err := b.isUserExists(username)
+	// если не существует или какая-то ошибка, выходим
+	if exists == nil || !*exists {
+		logger.Debug(fmt.Sprintf("Authorisation failed: user %s does not exist or err", username), op)
```
сделай метод Debugf, куда будет передавать строку для форматирования и аргументы, которые нужно вставить, чтобы fmt.Sprintf использовался под капотом, а не здесь

kanny4uHo commented on internal/auth/api/login_post.go:
```diff
@@ -0,0 +1,72 @@
+	// достаем инфу по пользователю по username
+	// создаем токен пользователя
+	tokenString, errM := b.AuthoriseUser(l.Username, l.Password)
+	if errM != nil {
```
в независимости от того, почему произошла ошибка, мы возвращаем 400. Надо разделить ответы на 400 и 500, в зависимости от того, успешно ли мы обработали запрос или упали по своей причине

kanny4uHo commented on internal/auth/api/models/init_valid_error_in_fields.go:
```diff
@@ -0,0 +1,19 @@
+// field string - название поле, в котором происходит валидация
+// op string - путь до функции, в которой происходит ошибка
+// формат op: service.dir_1.dir_2.MyValidateFunction
+func InitValidErrorInField(field string, op string) func(msg string) *er.MsgError {
```
NewValidErrorInField

kanny4uHo commented on internal/auth/behavior/behavior.go:
```diff
@@ -0,0 +1,173 @@
+
+	// сравниваем пароли
+	ok, errM := b.comparePassword(*user, password)
+	if errM != nil || !ok {
```
нужно разделить эти 2 случая, это 2 разных исхода

kanny4uHo commented on internal/auth/behavior/jwt/create_jwt.go:
```diff
@@ -0,0 +1,56 @@
+		Username: user.Username,
+		Role:     user.Role,
+		RegisteredClaims: jwt.RegisteredClaims{
+			ExpiresAt: jwt.NewNumericDate(time.Now().Add(time.Hour * (time.Duration(ttl)))), // Токен истекает через ttl часа
```
вокруг time.Duration можно сделать поменьше скобок

kanny4uHo commented on internal/auth/behavior/jwt/create_jwt.go:
```diff
@@ -0,0 +1,56 @@
+	// создаем структуру токена claims
+	claims := TokenClaims{
+		UserID:   user.UserID,
+		Username: user.Username,
```
давай тут хранить только userID? по нему будешь получать юзера потом

PtFux replied:
для создания профиля из кук при первом запросе

kanny4uHo replied:
расшифруй)

PtFux replied:
юзер зарегистрировался -> его редиректит на страницу профиля

к профилю приходит запрос от браузера, чтобы дал инфу по профилю. но у нас профиль и ауф - разные сервисы с собственной бд. скорее всего, в будущем, ауф - своя бд, профиль - ходит в общую, но это обсуждаемо.

ну и получается, что профиль не знает о регистрации пользователя. но ему приходит валидный токен в куки и он парсит оттуда инфу о пользователе и создает у себя в бдшке новый профиль под него

kanny4uHo replied:
профиль где-то возьмет данные сам, ему не нужно их подсовывать. Он сделает это или в соседнем сервисе, или соседний сервис будет приходить и класть профилю в бд эту информацию, откуда потом сам профиль будет брать инфу. Сейчас в любом случае профиль должен брать информацию из абстрактного репозитория, а что там будет - еще порешаем

kanny4uHo commented on cmd/microservices/auth/Dockerfile:
```diff
@@ -0,0 +1,62 @@
+RUN export $(grep -v '^#' /app/env | xargs)
+
+# Пробрасываем порт 8080 снаружи на 8081 внутри контейнера
+EXPOSE 8080
```
это так работает?) дырка проколота на 8080, а приложение работает на 8081

PtFux replied:
командой запуска пробрасываем порты

kanny4uHo replied:
теоретически в этот контейнер не сможет сходить другой контейнер, потому что приложение работает на неоткрытом порту

PtFux replied:
Ну можно везде 8080 написать

kanny4uHo commented on cmd/microservices/auth/Dockerfile:
```diff
@@ -0,0 +1,62 @@
+RUN mkdir -p /app/volume
+
+# Указываем том для монтирования
+VOLUME ["/app/volume"]
```
а что за volume?

PtFux replied:
на всякий случай presistance в докере)

kanny4uHo commented on internal/auth/api/utils/send_model.go:
```diff
@@ -0,0 +1,17 @@
+
+	// Сериализуем модель в JSON и отправляем в ответ
+	json.NewEncoder(w).Encode(model)
+	logger.StandardSendModel(model.String(), op)
```
мне кажется, если ты хочешь логировать вход и выход ручки, лучше использовать middleware

PtFux replied:
а есть миддллваре на выход?

kanny4uHo replied:
там вроде можно сделать  next.serve в коде миддлваре и после этого еще код писать

PtFux replied:
Справедливо

PtFux committed:
added mutex to repository

PtFux committed:
added tests

PtFux committed:
auth: status

PtFux committed:
added logging with f

PtFux committed:
auth: corrected special characters

PtFux committed:
auth: rename new validation error

PtFux committed:
auth: added явные Role consts

PtFux committed:
auth: removed '{'

kanny4uHo commented:
модуль buisness с опечаткой назван

PtFux committed:
auth: moved init behavior to handler struct

- создана структура хэндлеров
- behavior перенес в инит этой структуры
- переименован модуль business

PtFux committed:
added code in errors

PtFux committed:
auth: resolved problems

PtFux committed:
auth: resolved problems with conditional

PtFux committed:
auth: added smart errors

PtFux committed:
auth: small misstype

kanny4uHo commented:
многое порешали в личке в тг, пока в таком виде вливаем
